### PR TITLE
v0.20

### PR DESCRIPTION
**Pull request by vreuter**
I'd like to do a new version release in support of `looper`, see [this failed build](https://travis-ci.org/pepkit/looper/builds/495517975)

@nsheff do you have a way to quickly verify solutions for a few recent issues on `dev`?
- [x] submission script template field population, e.g. `{PARTITION}`
- [x] #253 
- [x] #258 
- [x] update reqs (sphinx, check test package version bounds; `attmap`)
- [x] https://github.com/pepkit/attmap/pull/11
- [x] https://github.com/pepkit/divvy/pull/12
- [x] confirm functional with `caravel`
- [x] confirm functional with both release and dev versions of `looper`

### CODE REVIEW

**Comment by vreuter** [comment]
Thanks @michalstolarczyk I'll look into this. I think the annotation path should be updated according to the activated subproject.

**Comment by vreuter** [comment]
@nsheff do you have a way to quickly verify solutions for a few recent issues on `dev`?
- [ ] submission script template field population, e.g. `{PARTITION}`
- [ ] #253 
- [ ] #258

**Comment by nsheff** [comment]
> @nsheff do you have a way to quickly verify solutions for a few recent issues on `dev`?

Yes, I'm in the process of running the AML samples through the meth pipeline; that is what I have been using. I can use that to verify. Let me know when you're at a point that you want me to do so.

**Comment by nsheff** [comment]
@vreuter I had assigned a few issues to the 0.2 milestone: https://github.com/pepkit/peppy/milestone/11

did you want to do a release before those or do you think it's reasonable to batch all that into this release? Those seemed like a bunch of related things to me.

**Comment by nsheff** [comment]
@michalstolarczyk might be able to contribute to the divvy issues assigned to that milestone: 

https://github.com/pepkit/peppy/issues/244
https://github.com/pepkit/peppy/issues/237

we could do a few days blitz on peppy and knock these all out.

Also we may be introducing backwards-incompatible changes here so maybe we should consider updating to a major version change (like v1.0)? But maybe we want to wait until snakemake integration is complete... ahh, feature creep...

**Comment by michalstolarczyk** [comment]
> @michalstolarczyk might be able to contribute to the divvy issues assigned to that milestone:

Sure, I can try to resolve these

> snakemake integration

wow, didn't know about it!

**Comment by vreuter** [comment]
For the next release I'd like to do all but the `snakemake` integration but I don't have a strong versioning preference. It feels nice to make that integration part of the 0 --> 1 switch, but that nicety feels like it's worth neither the feature creep like you mentioned (i.e., waiting to release more critical changes and making the next release 1.0) nor the clear communication of the nature of the changes (perhaps sacrificed if we make it 0.20 rather than 1.0). Again though, I feel more strongly about releasing these changes before Snakemake than I do about the choice of version number.

**Comment by nsheff** [comment]
ok, let's go for the current .20 milestone as [currently specified](https://github.com/pepkit/peppy/milestone/11). The target release date is next week. Not sure it will take that long though, honestly, much of this stuff is done.

**Comment by michalstolarczyk** [comment]
Do we want `divvy` to start using `attmap` (https://github.com/databio/divvy/issues/8) before this merge?

**Comment by nsheff** [comment]
> Do we want `divvy` to start using `attmap` ([databio/divvy#8](https://github.com/databio/divvy/issues/8)) before this merge?

Yes, let's do it.

**Comment by vreuter** [comment]
Something to note for examples...one provided above features `library`...I couldn't find those file(s) tracked anywhere (though I only checked a couple projects and branches), but if we provide those as public-facing examples they should use `protocol` rather than `library` so that someone running an example doesn't hit a deprecation warning.

**Comment by nsheff** [comment]
yep. library was the old term, before protocol, from several years ago.

**Comment by vreuter** [comment]
> yep. library was the old term, before protocol, from several years ago.

Right, I wasn't asking why it was there or where it came from, just noting that we shouldn't be making or introducing new examples with it.

File: requirements/requirements-all.txt [hunk 1]
```diff
@@ -1,4 +1,5 @@
 attmap>=0.1.8
 pandas>=0.20.2
 pyyaml>=3.12
+divvy
```
Comment (nsheff) [inline requirements/requirements-all.txt]: we will need to release divvy to pypi I guess before this.

**Comment by nsheff** [comment]
@vreuter I'm trying to build the peppy docs with the latest version of oradocle and mkdocs-databio but it's not working... I merged in your branch as it originally let me build, but once I changed the  `autodoc_package` from `divvy` (it was set incorrectly) to `peppy` it's throwing errors...

see if you can build this now and make sure all your external changes are uptodate on those other packages?

**Comment by vreuter** [comment]
Sorry, the merge should've been `autodoc` --> `mkdocs` --> `dev`, so that may be causing issues

**Comment by vreuter** [comment]
I think it's just a lots-of-branches / projects issue...it seems to be working with `autodoc` branch of `peppy` installed, along with `autodoc` branch of the MkDocs plugin installed, using `master` of `oradocle`.

**Comment by vreuter** [comment]
@nsheff https://github.com/pepkit/peppy/pull/267 is open now and should address the docs build issue. If that build passes, it should be ready for merge to `dev` and then this PR will once again be "ready."

**Comment by nsheff** [comment]
@vreuter from a docs perspective I'm happy with this;

@michalstolarczyk can you confirm that this works for `caravel v0.2` ?

@vreuter can you confirm that this works for both the release and dev/argparse versions of looper?

once that's confirmed please release.

**Comment by nsheff** [comment]
is `make_docs.sh` still necessary?

**Comment by michalstolarczyk** [comment]
> @michalstolarczyk can you confirm that this works for `caravel v0.2` ?

works

**Comment by vreuter** [comment]
> is make_docs.sh still necessary?

I don't know what this did.

**Comment by vreuter** [comment]
> @vreuter can you confirm that this works for both the release and dev/argparse versions of looper?

Sure, how should this be confirmed?

**Comment by nsheff** [comment]
Ok I'm testing it... I get this error with the current dev version of looper:

```
looper --version
Traceback (most recent call last):
  File "/home/ns5bc/.local/bin/looper", line 7, in <module>
    from looper.looper import main
  File "/home/ns5bc/.local/lib/python2.7/site-packages/looper/__init__.py", line 22, in <module>
    from peppy import \
ImportError: cannot import name COMPUTE_SETTINGS_VARNAME
```

```
import peppy

In [2]: peppy.__version__
Out[2]: '0.20'
import looper
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
<ipython-input-3-7360e9cb3c09> in <module>()
----> 1 import looper

/home/ns5bc/.local/lib/python2.7/site-packages/looper/__init__.py in <module>()
     20 # the locations of some of the peppy declarations. Effectively, concentrate
     21 # the connection between peppy and looper here, to the extent possible.
---> 22 from peppy import \
     23     FLAGS, IMPLICATIONS_DECLARATION, SAMPLE_INDEPENDENT_PROJECT_SECTIONS, \
     24     SAMPLE_NAME_COLNAME, COMPUTE_SETTINGS_VARNAME

ImportError: cannot import name COMPUTE_SETTINGS_VARNAME
```

**Comment by nsheff** [comment]
with the current release version of looper (`v.10`) I get this error:

```
looper --version
Traceback (most recent call last):
  File "/home/ns5bc/.local/bin/looper", line 7, in <module>
    from looper.looper import main
  File "/home/ns5bc/.local/lib/python2.7/site-packages/looper/__init__.py", line 14, in <module>
    from .pipeline_interface import PipelineInterface
  File "/home/ns5bc/.local/lib/python2.7/site-packages/looper/pipeline_interface.py", line 17, in <module>
    from peppy import utils, AttributeDict, Sample
ImportError: cannot import name AttributeDict
```

**Comment by nsheff** [comment]
Ok so the question is this: are we intentionally planning to break backwards compatibility so this version of `peppy` will not work with any previous versions of `looper`? this the point of no return.

In this case, this really needs to move to `peppy v1.0`, because this is a major refactoring. I was hoping this would be at least *mostly* compatible with earlier versions.

**Comment by michalstolarczyk** [comment]
It works with looper on `argparser_caravel`. I thought we consider this one it the "development" version of looper. 

The `divvy` related changes were pushed there and the error on `dev` looks like it's caused by these.

**Comment by nsheff** [comment]
Ok, **confirmed**. that's fine -- 

The only problem then is that peppy 0.2 is **incompatible** with looper 0.10.

Would it be simple enough to preserve this compatibility so we don't have to introduce major version changes for both peppy and looper yet?

**Comment by vreuter** [comment]
@nsheff can you confirm the desired `import` behavior on previous `looper` version(s)? If that succeeds, is this good for merge (i.e., is the final remaining tick-box in the description of this PR "done" given the success of the RRBS sample(s) test case run earlier with updated `looper`)?

**Comment by nsheff** [comment]
yep -- working on it right now:

does this look easy to solve?

```
Traceback (most recent call last):
  File "/home/ns5bc/.local/bin/looper", line 11, in <module>
    load_entry_point('looper==0.10', 'console_scripts', 'looper')()
  File "/home/ns5bc/.local/lib/python2.7/site-packages/looper/looper.py", line 838, in main
    compute_env_file=getattr(args, 'env', None))
  File "/home/ns5bc/.local/lib/python2.7/site-packages/looper/project.py", line 44, in __init__
    process_pipeline_interfaces(self.metadata.pipelines_dir)
  File "/home/ns5bc/.local/lib/python2.7/site-packages/looper/project.py", line 236, in process_pipeline_interfaces
    pipe_iface = PipelineInterface(pipe_iface_location)
  File "/home/ns5bc/.local/lib/python2.7/site-packages/looper/pipeline_interface.py", line 45, in __init__
    super(PipelineInterface, self).__init__(_attribute_identity=False)
TypeError: __init__() got an unexpected keyword argument '_attribute_identity'
```

**Comment by vreuter** [comment]
No. I thought you just wanted `--version` to work. There are way too many changes to make actually running things work. The user should be able to check their version I think but then should be forced to upgrade.

**Comment by nsheff** [comment]
well, that's a minimum. I was hoping it would work :). But ok, let's go with this.

But, I think we should wait to release this to do it simultaneously with `looper v0.11`. -- don't you agree? otherwise we're sitting with incompatible master released versions until that release...

**Comment by vreuter** [comment]
> But, I think we should wait to release this to do it simultaneously with `looper v0.11`. -- don't you agree? otherwise we're sitting with incompatible master released versions until that release...

Absolutely -- forgot that we have some more time on `looper`

**Comment by vreuter** [comment]
Wait for https://github.com/pepkit/looper/pull/109 and release `peppy` and `looper` simultaneously per @nsheff

**Comment by vreuter** [comment]
Wait for `looper v0.11` release

**Comment by vreuter** [comment]
Wait until `looper v0.11` release is ready

**Comment by vreuter** [comment]
Tests here and in `looper` are fine. @nsheff @michalstolarczyk pending confirmation of the tickboxes in the main comment for this PR in whatever way seems good, this can be merged.

**Comment by michalstolarczyk** [comment]
works with `caravel`

**Comment by nsheff** [comment]
logmuse will have to be on pypi before we can release

**Comment by nsheff** [comment]
alright, ready to release this -- any objections? speak now...

**Comment by vreuter** [comment]
> alright, ready to release this -- any objections? speak now...

None from me

**Comment by vreuter** [comment]
:crossed_fingers:

**Comment by nsheff** [comment]
@vreuter can you go through these and either close or change milestone so we can close the 0.20 milestone?

https://github.com/pepkit/peppy/milestone/11

**Comment by vreuter** [comment]
Sure

**Pull request merged**